pkg/detectors/airtableapikey/airtableapikey.go

```diff
@@ -45,65 +48,109 @@ type response struct {
-
-	if len(keyMatches) == 0 {
-		keyMatches = personalKeyMatches
+	appMatches := make(map[string]struct{})
+	for _, matches := range appPat.FindAllStringSubmatch(dataStr, -1) {
+		appMatches[matches[1]] = struct{}{}
+	}
+	keyMatches := make(map[string]struct{})
+	for _, matches := range keyPat.FindAllStringSubmatch(dataStr, -1) {
+		keyMatches[matches[1]] = struct{}{}
+	}
+	for _, matches := range personalPat.FindAllStringSubmatch(dataStr, -1) {
+		keyMatches[matches[1]] = struct{}{}
 	}
 
-	for _, keyMatch := range keyMatches {
-		if len(keyMatch) != 2 {
-			continue
-		}
-
-		keyRes := strings.TrimSpace(keyMatch[1])
-
-		for _, appMatch := range appMatches {
-			if len(appMatch) != 2 {
-				continue
-			}
-			appRes := strings.TrimSpace(appMatch[1])
+	for keyMatch := range keyMatches {
+		var (
+			r        *detectors.Result
+			appMatch string
+		)
 
-			s1 := detectors.Result{
-				DetectorType: detectorspb.DetectorType_AirtableApiKey,
-				Redacted:     appRes,
-				Raw:          []byte(keyRes),
-				RawV2:        []byte(keyRes + appRes),
-			}
+		for a := range appMatches {
+			appMatch = a
 
 			if verify {
-				req, err := http.NewRequestWithContext(ctx, "GET", "https://api.airtable.com/v0/"+appRes+"/Projects", nil)
-				if err != nil {
-					continue
+				client := s.client
+				if client == nil {
+					client = defaultClient
 				}
-				req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", keyRes))
-				res, err := client.Do(req)
-				if err == nil {
-					defer res.Body.Close()
-					if res.StatusCode >= 200 && res.StatusCode < 300 {
-						s1.Verified = true
-					} else if res.StatusCode == 403 {
-						var resp response
-						if err = json.NewDecoder(res.Body).Decode(&resp); err == nil {
-							// check if the error is due to invalid permissions or model not found
-							if resp.Error.Type == "INVALID_PERMISSIONS_OR_MODEL_NOT_FOUND" {
-								// The key is verified as it works, but the user must enumerate the tables or permissions for the key.
-								s1.Verified = true
-							}
-						}
-					}
+
+				isVerified, verificationErr := verifyMatch(ctx, client, appMatch, keyMatch)
+				if isVerified {
+					r = createResult(appMatch, keyMatch, isVerified, verificationErr)
+					break
 				}
 			}
+		}
 
-			results = append(results, s1)
+		if r == nil {
+			if len(appMatches) != 1 {
+				appMatch = ""
+			}
+			r = createResult(appMatch, keyMatch, false, nil)
 		}
+		results = append(results, *r)
 	}
 
 	return results, nil
 }
 
+func createResult(app string, key string, verified bool, err error) *detectors.Result {
+	r := &detectors.Result{
+		DetectorType: detectorspb.DetectorType_AirtableApiKey,
+		Raw:          []byte(key),
+		Redacted:     app,
+		Verified:     verified,
+	}
+
+	if app != "" {
+		r.RawV2 = []byte(fmt.Sprintf(`%s:%s`, app, key))
+	}
+
+	if err != nil {
+		r.SetVerificationError(err, key)
+	}
+	return r
+}
+
+func verifyMatch(ctx context.Context, client *http.Client, app string, key string) (bool, error) {
```

Comment:
I think we should refactor detectors to follow a consistent structure. Adding this function to all detectors would help isolate the verification logic, making the code more organized and easier to read. Many already have it but some don't.

Reply:
I think each detector needs separate `Detect` and `Verify` functions. Then the engine could cache results and only attempt to verify new ones.